pkg/apis/node/types.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to select or configure the container runtime that is
+// used to run all containers in a pod. RuntimeClasses are (currently) manually
+// defined by a user or cluster provisioner, and referenced in the PodSpec. The
+// Kubelet is responsible for resolving the RuntimeClassName reference before
+// running the pod.  For more details, see
+// https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md
+type RuntimeClass struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that required to describe the RuntimeClass to the Container Runtime Interface
+// (CRI) implementation, as well as any other components that need to understand
+// how the pod will be run. The RuntimeClassSpec should be treated as immutable.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
```

Comment:
All these terms are so overloaded. I like "containerizers", but my concerns are that no one will know what that means (not that anyone knows what runtime means either...). The field name is called "RuntimeHandler", so we're already locked into that to some degree (unless we want to rename it for beta? Although it's also called runtime_handler in the CRI).

Another possibility would be to say "OCI runtime" here, since that's what it is in practice. However, there's nothing to say I couldn't select between different non-OCI runtimes, or just different configurations of the CRI implementation.

Open to suggestions here...

Reply:
Does CRI "officially" support non-OCI runtimes? 
One alternative is to not use the word runtime and instead say this field is a config passed to CRI "runtimes" to choose appropriate "sandboxing" solutions. (Note "sandbox" is a new term in this case :( )

Reply:
There are a couple of mentions of OCI in the [CRI spec](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto), but AFAIK it's not required. For instance, although gvisor & kata both support the OCI interface, they're working on integrating with containerd at the higher-level [contaired shim v2](https://github.com/containerd/containerd/issues/2426) API.

Reply:
This doesn't select between CRI implementations, right?  It passes info *through* CRI (ostensibly to "the runtime") to select a set of behaviors or sub-runtimes.  Am I getting that right?

I don't hate "runtime" but I do think it's worth contemplating

Reply:
@thockin correct. This is essentially an opaque value from Kubernetes perspective. We pass it off to the CRI, and it is completely up to the CRI for what to do with it:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L365-L370

For both containerd & CRI-O, this is implemented as a key that selects from different OCI (or shim v2) level configurations.

An alternative approach to runtime_handler would be to just key off the RuntimeClass name directly, but I wanted to decouple the handler selection from the RuntimeClass resource. For instance, if an operator wanted to create a new RuntimeClass to handle rolling updates (as discussed [here](https://github.com/kubernetes/enhancements/pull/723)).

Reply:
conclusion from offline discussion: if possible in CRD -> built in conversion, rename to `handler`